File: apps/web/src/api/context.ts
import type { ConversationFetch } from "./conversations";
import { MessageApiError } from "./messages";

export type WorkspaceContext = {
  environment: "development" | "synthetic";
  device: string | null;
  lot: string | null;
  time_range: string | null;
  data_source: "synthetic_demo";
};
export type WorkspaceContextUpdate = Pick<WorkspaceContext, "device" | "lot" | "time_range">;

const isContext = (value: unknown): value is WorkspaceContext => {
  if (typeof value !== "object" || value === null) return false;
  const context = value as WorkspaceContext;
  return (context.environment === "development" || context.environment === "synthetic") &&
    (context.device === null || typeof context.device === "string") &&
    (context.lot === null || typeof context.lot === "string") &&
    (context.time_range === null || typeof context.time_range === "string") &&
    context.data_source === "synthetic_demo";
};

export function buildContextUrl(conversationId: string, baseUrl?: string): string {
  const base = baseUrl?.trim() ? baseUrl.replace(/\/+$/, "") : "/api";
  return `${base}/conversations/${conversationId}/context`;
}

export async function getWorkspaceContext(conversationId: string, fetchImplementation: ConversationFetch = fetch): Promise<WorkspaceContext> {
  try {
    const response = await fetchImplementation(buildContextUrl(conversationId, import.meta.env.VITE_API_BASE_URL), { headers: { Accept: "application/json" } });
    if (!response.ok) throw new Error("unsuccessful response");
    const payload: unknown = await response.json();
    if (!isContext(payload)) throw new Error("invalid response");
    return payload;
  } catch (error) {
    throw new MessageApiError(error);
  }
}

export async function updateWorkspaceContext(conversationId: string, update: WorkspaceContextUpdate, fetchImplementation: ConversationFetch = fetch): Promise<WorkspaceContext> {
  try {
    const response = await fetchImplementation(buildContextUrl(conversationId, import.meta.env.VITE_API_BASE_URL), { method: "PATCH", headers: { Accept: "application/json", "Content-Type": "application/json" }, body: JSON.stringify(update) });
    if (!response.ok) throw new Error("unsuccessful response");
    const payload: unknown = await response.json();
    if (!isContext(payload)) throw new Error("invalid response");
    return payload;
  } catch (error) { throw new MessageApiError(error); }
}

File: apps/api/src/industrial_agent/schemas/context.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Environment = Literal["development", "synthetic"]
DataSource = Literal["synthetic_demo"]


class WorkspaceContextUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    environment: Environment | None = None
    device: str | None = Field(default=None, max_length=200)
    lot: str | None = Field(default=None, max_length=200)
    time_range: str | None = Field(default=None, max_length=100)
    data_source: DataSource | None = None


class WorkspaceContextRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    environment: Environment
    device: str | None
    lot: str | None
    time_range: str | None
    data_source: DataSource
